relay.go

```diff
@@ -52,166 +61,271 @@ func NewRelay(ctx context.Context, h host.Host, opts ...RelayOpt) (*Relay, error
+func (r *Relay) handleNewStream(s inet.Stream) {
+	log.Infof("new relay stream from: %s", s.Conn().RemotePeer())
+
+	rd := newDelimitedReader(s, maxMessageSize)
+
+	var msg pb.CircuitRelay
+
+	err := rd.ReadMsg(&msg)
 	if err != nil {
-		return &RelayStatus{
-			Code:    StatusDstAddrErr,
-			Message: err.Error(),
-		}
+		r.handleError(s, pb.CircuitRelay_MALFORMED_MESSAGE)
+		return
 	}
 
-	log.Infof("relay connection from: %s", info.ID)
-	select {
-	case r.incoming <- &Conn{Stream: s, remoteMaddr: info.Addrs[0], remotePeer: info.ID}:
-		return nil
-	case <-time.After(RelayAcceptTimeout):
-		return &RelayStatus{
-			Code:    StatusDstRelayRefused,
-			Message: "timed out waiting for relay to be accepted",
-		}
+	switch msg.GetType() {
+	case pb.CircuitRelay_HOP:
+		r.handleHopStream(s, &msg)
+	case pb.CircuitRelay_STOP:
+		r.handleStopStream(s, &msg)
+	case pb.CircuitRelay_CAN_HOP:
+		r.handleCanHop(s, &msg)
+	default:
+		log.Warningf("unexpected relay handshake: %d", msg.GetType())
+		r.handleError(s, pb.CircuitRelay_MALFORMED_MESSAGE)
 	}
 }
 
-func (r *Relay) HandleNewHopStream(s inet.Stream) {
-	log.Infof("new hop stream from: %s", s.Conn().RemotePeer())
-	status := r.handleNewHopStream(s)
-	if status != nil {
-		if err := status.WriteTo(s); err != nil {
-			log.Debugf("problem writing error status back: %s", err)
-			s.Close()
-			return
-		}
+func (r *Relay) handleHopStream(s inet.Stream, msg *pb.CircuitRelay) {
+	if !r.hop {
+		r.handleError(s, pb.CircuitRelay_HOP_CANT_SPEAK_RELAY)
+		return
 	}
-}
 
-func (r *Relay) handleNewHopStream(s inet.Stream) *RelayStatus {
-	info, err := r.readInfo(s)
+	src, err := peerToPeerInfo(msg.GetSrcPeer())
 	if err != nil {
-		return &RelayStatus{
-			Code:    StatusRelayAddrErr,
-			Message: err.Error(),
-		}
+		r.handleError(s, pb.CircuitRelay_HOP_SRC_MULTIADDR_INVALID)
+		return
 	}
 
-	if info.ID == r.self {
-		return &RelayStatus{
-			Code:    StatusRelayHopToSelf,
-			Message: "relay hop attempted to self",
-		}
+	if src.ID != s.Conn().RemotePeer() {
+		r.handleError(s, pb.CircuitRelay_HOP_SRC_MULTIADDR_INVALID)
+		return
+	}
+
+	dst, err := peerToPeerInfo(msg.GetDstPeer())
+	if err != nil {
+		r.handleError(s, pb.CircuitRelay_HOP_DST_MULTIADDR_INVALID)
+		return
+	}
+
+	if dst.ID == r.self {
+		r.handleError(s, pb.CircuitRelay_HOP_CANT_RELAY_TO_SELF)
+		return
 	}
 
-	ctp := r.host.Network().ConnsToPeer(info.ID)
+	// open stream
+	ctp := r.host.Network().ConnsToPeer(dst.ID)
 	if len(ctp) == 0 {
-		return &RelayStatus{
-			Code:    StatusRelayNotConnected,
-			Message: "refusing to make new connection for relay",
+		if !r.active {
+			r.handleError(s, pb.CircuitRelay_HOP_NO_CONN_TO_DST)
+			return
+		}
+
+		ctx, cancel := context.WithTimeout(r.ctx, time.Second*10)
+		defer cancel()
+		err = r.host.Connect(ctx, dst)
+		if err != nil {
+			log.Debugf("error opening relay connection to %s: %s", dst.ID.Pretty(), err.Error())
+			r.handleError(s, pb.CircuitRelay_HOP_CANT_DIAL_DST)
+			return
```

Comment:
See above for unawareness of implicit connect on new stream construction. 
On the other hand, we want to distinguish Dial errors from Stream opening errors (there are separate error codes and helps debugging), so I think we want to keep the explicit connect.